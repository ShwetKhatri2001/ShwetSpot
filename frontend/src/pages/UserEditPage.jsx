import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import { Button, Form } from 'react-bootstrap';
import { useDispatch, useSelector } from 'react-redux';
import Message from '../components/Message';
import Loader from '../components/Loader';
import { getUserDetails, updateUser } from '../redux/actions/userActions';
import { USER_UPDATE_RESET } from '../redux/constants/userConstants';

const UserEditPage = ({ match, history }) => {
  const userId = match.params.id;

  const [name, setName] = useState('');
  const [email, setEmail] = useState('');
  const [isAdmin, setIsAdmin] = useState(false);

  const dispatch = useDispatch();

  const userDetails = useSelector(state => state.userDetails);
  const { loading, error, user } = userDetails;

  const userUpdate = useSelector(state => state.userUpdate);
  const {
    loading: loadingUpdate,
    error: errorUpdate,
    success: successUpdate,
  } = userUpdate;

  useEffect(() => {
    if (successUpdate) {
      dispatch(getUserDetails('profile'));
      dispatch({ type: USER_UPDATE_RESET });
      history.push('/admin/userlist');
    } else {
      if (!user.name || user._id !== userId) {
        dispatch(getUserDetails(userId));
      } else {
        setName(user.name);
        setEmail(user.email);
        setIsAdmin(user.isAdmin);
      }
    }
  }, [dispatch, history, userId, user, successUpdate]);

  const submitHandler = e => {
    e.preventDefault();
    dispatch(updateUser({ _id: userId, name, email, isAdmin }));
  };

  return (
    <>
      <Link
        to="/admin/userlist"
        className="btn btn-danger my-3"
        onClick={() => dispatch(getUserDetails('profile'))}
      >
        Go Back
      </Link>
      <div className="justify-content-center align-items-center container bg-light p-4">
        <h1>Edit User</h1>
        {loadingUpdate && <Loader />}
        {errorUpdate && <Message variant="danger">{errorUpdate}</Message>}
        {loading ? (
          <Loader />
        ) : error ? (
          <Message variant="danger">{error}</Message>
        ) : (
          <Form onSubmit={submitHandler}>
            <Form.Group controlId="name">
              <Form.Label>Name</Form.Label>
              <Form.Control
                type="name"
                placeholder="Enter name"
                value={name}
                onChange={e => setName(e.target.value)}
              />
            </Form.Group>

            <Form.Group controlId="email">
              <Form.Label>Email Address</Form.Label>
              <Form.Control
                type="email"
                placeholder="Enter email"
                value={email}
                onChange={e => setEmail(e.target.value)}
              />
            </Form.Group>

            <Form.Group controlId="isadmin">
              <Form.Check
                type="checkbox"
                label="Is Admin"
                checked={isAdmin}
                onChange={e => setIsAdmin(e.target.checked)}
              />
            </Form.Group>

            <Button type="submit" variant="success">
              Update
            </Button>
          </Form>
        )}
      </div>
    </>
  );
};

export default UserEditPage;
