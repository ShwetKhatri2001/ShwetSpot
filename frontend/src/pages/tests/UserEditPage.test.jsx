import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

import UserEditPage from '../UserEditPage';
import store from '../../redux/store';

describe('UserEditPage Component', () => {
  const MockUserEditPage = (
    <MemoryRouter>
      <Provider store={store}>
        <UserEditPage match={{ params: { id: 1 } }} />
      </Provider>
    </MemoryRouter>
  );

  it('should match the snapshot', () => {
    const tree = renderer.create(MockUserEditPage).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
