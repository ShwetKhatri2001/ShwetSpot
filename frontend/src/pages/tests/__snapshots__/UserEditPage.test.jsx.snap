// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserEditPage Component should match the snapshot 1`] = `
Array [
  <a
    className="btn btn-danger my-3"
    href="/admin/userlist"
    onClick={[Function]}
  >
    Go Back
  </a>,
  <div
    className="justify-content-center align-items-center container bg-light p-4"
  >
    <h1>
      Edit User
    </h1>
    <form
      className=""
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <label
          className="form-label"
          htmlFor="name"
        >
          Name
        </label>
        <input
          className="form-control"
          id="name"
          onChange={[Function]}
          placeholder="Enter name"
          type="name"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <label
          className="form-label"
          htmlFor="email"
        >
          Email Address
        </label>
        <input
          className="form-control"
          id="email"
          onChange={[Function]}
          placeholder="Enter email"
          type="email"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <div
          className="form-check"
        >
          <input
            checked={false}
            className="form-check-input"
            disabled={false}
            id="isadmin"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="form-check-label"
            htmlFor="isadmin"
            title=""
          >
            Is Admin
          </label>
        </div>
      </div>
      <button
        className="btn btn-success"
        disabled={false}
        type="submit"
      >
        Update
      </button>
    </form>
  </div>,
]
`;
